docs/features/source-generators.cookbook.md

```diff
@@ -624,105 +624,83 @@ public class MyGenerator : ISourceGenerator
+                solution = solution.WithProjectCompilationOptions(projectId, compilationOptions);
 
-            // We can now assert things about the resulting compilation:
-            Debug.Assert(diagnostics.IsEmpty); // there were no diagnostics created by the generators
-            Debug.Assert(outputCompilation.SyntaxTrees.Count() == 2); // we have two syntax trees, the original 'user' provided one, and the one added by the generator
-            Debug.Assert(outputCompilation.GetDiagnostics().IsEmpty); // verify the compilation with the added source has no diagnostics
+                return solution;
+            });
+        }
 
-            // Or we can look at the results directly:
-            GeneratorDriverRunResult runResult = driver.GetRunResult();
+        public LanguageVersion LanguageVersion { get; set; } = LanguageVersion.Default;
 
-            // The runResult contains the combined results of all generators passed to the driver
-            Debug.Assert(runResult.GeneratedTrees.Length == 1);
-            Debug.Assert(runResult.Diagnostics.IsEmpty);
+        private static ImmutableDictionary<string, ReportDiagnostic> GetNullableWarningsFromCompiler()
+        {
+            string[] args = { "/warnaserror:nullable" };
+            var commandLineArguments = CSharpCommandLineParser.Default.Parse(args, baseDirectory: Environment.CurrentDirectory, sdkDirectory: Environment.CurrentDirectory);
+            var nullableWarnings = commandLineArguments.CompilationOptions.SpecificDiagnosticOptions;
 
-            // Or you can access the individual results on a by-generator basis
-            GeneratorRunResult generatorResult = runResult.Results[0];
-            Debug.Assert(generatorResult.Generator == generator);
-            Debug.Assert(generatorResult.Diagnostics.IsEmpty);
-            Debug.Assert(generatorResult.GeneratedSources.Length == 1);
-            Debug.Assert(generatorResult.Exception is null);
+            return nullableWarnings;
         }
-
-        private static Compilation CreateCompilation(string source)
-            => CSharpCompilation.Create("compilation",
-                new[] { CSharpSyntaxTree.ParseText(source) },
-                new[] { MetadataReference.CreateFromFile(typeof(Binder).GetTypeInfo().Assembly.Location) },
-                new CSharpCompilationOptions(OutputKind.ConsoleApplication));
     }
 }
 ```
 
-Note: the above example uses MSTest, but the contents of the test are easily adapted to other frameworks, such as XUnit.
+Then, in your test file:
+
+```csharp
+using VerifyCS = CSharpSourceGeneratorVerifier<YourGenerator>;
+```
+
+And use the following in your test method:
+
+```csharp
+var code = "initial code"
+var generated = "expected generated code";
+await new VerifyCS.Test
+{
+    TestState = 
+    {
+        Sources = { code },
+    },
+    FixedState = 
+    {
+        Sources = { code, generated },
+    }
+}.RunAsync();
```

Comment:
This was rewritten in https://github.com/dotnet/roslyn-analyzers/commit/8b0215e8b67e4d10695ad7d2a0cfd7e210cad869